elastic/kibana
# [WIP] Move mapping-building logic into the migrations package

_This is one in a series of several PRs that will go towards resolving #15100._

This PR moves the logic for building and validating mappings into the migrations package. Ultimately, this is functionality that will be owned by the migrations system, and this is a small step in that direction. This change allows us to more easily build mappings for indices other than `.kibana`. It also allows us to eventually build / process mapping information earlier in the Kibana boot process.

## Testing this PR

This change introduces a handful of changes:

- Moves the logic for building mappings and updating the index template / mappings.
- No longer validates that the index has a single root level document, but instead assumes 'doc' as the root
- Adds the ability to add mappings to an existing mapping
- Throws an error if a destructive mapping change is detected

One scenario to check specifically is:

- Start Kibana w/ some plugins enabled
- Write data using those plugins
- Restart Kibana w/ one or more of those plugins disabled
- Check that:
  - Kibana boots fine
  - Saved object writes still work
